### PR TITLE
Tap tests

### PR DESCRIPTION
**Pull request by ababino**
Test governance and tap behavior. 
@facuspagnuolo

### CODE REVIEW

File: contracts/DAICO.sol [hunk 1]
```diff
@@ -112,6 +112,25 @@ contract DAICO is Crowdsale, Ownable {
...
+        * there are more votes in favor than aginst the proposal, and the number
+        * of votes is greater than the quorum, then, the tap variable is updated
+        * to the propsed new tap.
+        * @param proposalID The ID of the proposal to execute.
+        */
+        function executeRaiseTapProposal(uint256 proposalID) public {
+          RaiseTapProposal storage proposal = proposals[proposalID];
+          require(proposal.votingDeadline < block.timestamp);
+          require(!proposal.executed);
+          require(proposal.proposedNewTap > tap);
```
Comment (facuspagnuolo) [inline contracts/DAICO.sol]: just asking, would it make sense decreasing the tap in a particular scenario?

File: test/DAICO.test.js [hunk 2]
```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
...
-  describe('voting as expected', function () {
+  describe('proposals', function () {
     it('no purchaser no holder', async function() {
       let purchaser = accounts[5];
       let isholder = await this.daico.isHolder(purchaser);
       assert.ok(!isholder);
     });
 
-    it('if yoy did not purchased a token you are not a holder', async function() {
+    it('if you did not purchased a token you are not a holder', async function() {
```
Comment (facuspagnuolo) [inline test/DAICO.test.js]: should be `purchasers are considered holders` right?

File: test/DAICO.test.js [hunk 3]
```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
...
       let purchaser = accounts[4];
       await this.daico.buyTokens(purchaser, {from: purchaser, value: higher_value}).should.be.fulfilled;
       let isholder = await this.daico.isHolder(purchaser);
       assert.ok(isholder);
     });
 
-    it('holder can propose a new tap', async function() {
+    it('holder can propose a new tap in tap mode', async function() {
```
Comment (facuspagnuolo) [inline test/DAICO.test.js]: what do you mean with `tap mode`?

File: test/DAICO.test.js [hunk 4]
```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
...
 
-    it('you can not propose lower the tap', async function() {
+    it('holder can vote a proposal in time', async function() {
       let purchaser = accounts[4];
-      await this.daico.buyTokens(purchaser, {from: purchaser, value: higher_value}).should.be.fulfilled;
-      await this.daico.newRaiseTapProposal(100, 3600, {from: purchaser}).should.be.rejectedWith(EVMRevert);
+      await this.daico.buyTokens(purchaser, {from: purchaser, value: higher_value});
+      await increaseTimeTo(this.afterlastWithdrawn);
+      await this.daico.newRaiseTapProposal(100, 3600, {from: purchaser}).should.be.fulfilled;
+      await this.daico.vote(0, true, {from: purchaser}).should.be.fulfilled;
```
Comment (facuspagnuolo) [inline test/DAICO.test.js]: we need to check whatever you're expecting to happen from the contract side after voting

File: test/DAICO.test.js [hunk 5]
```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
...
+      await this.daico.buyTokens(purchaser, {from: purchaser, value: higher_value});
+      await increaseTimeTo(this.afterlastWithdrawn);
+      const proposalID = await this.daico.newRaiseTapProposal(100, 3600, {from: purchaser});
+      await this.daico.vote(0, false, {from: purchaser}).should.be.fulfilled;;
+      await increaseTimeTo(latestTime() + duration.hours(1) + duration.seconds(1));
+      let old_tap = await this.daico.tap();
+      assert.ok(old_tap.c[0]==0);
+      await this.daico.executeRaiseTapProposal(0);
+      let new_tap = await this.daico.tap();
+      assert.ok(new_tap.c[0]==0);
```
Comment (facuspagnuolo) [inline test/DAICO.test.js]: use bignumber assertions

File: test/DAICO.test.js [hunk 6]
```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
...
+      await this.daico.buyTokens(purchaser, {from: purchaser, value: higher_value});
+      await increaseTimeTo(this.afterlastWithdrawn);
+      const proposalID = await this.daico.newRaiseTapProposal(100, 3600, {from: purchaser});
+      await this.daico.vote(0, true, {from: purchaser}).should.be.fulfilled;;
+      await increaseTimeTo(latestTime() + duration.hours(1) + duration.seconds(1));
+      let old_tap = await this.daico.tap();
+      assert.ok(old_tap.c[0]==0);
+      await this.daico.executeRaiseTapProposal(0);
+      let new_tap = await this.daico.tap();
+      assert.ok(new_tap.c[0]==100);
```
Comment (facuspagnuolo) [inline test/DAICO.test.js]: we should check the proposal has been updated as we expected

File: test/DAICO.test.js [hunk 7]
```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
...
+      await this.daico.newRaiseTapProposal(tap, 3600, {from: no_owner});
+      await this.daico.vote(0, true, {from: no_owner}).should.be.fulfilled;;
+      await increaseTimeTo(latestTime() + duration.hours(1) + duration.seconds(1));
+      await this.daico.executeRaiseTapProposal(0).should.be.fulfilled;
+      await increaseTimeTo(latestTime() + duration.seconds(1000000));
+      const initial_balance = web3.eth.getBalance(owner).c[0];
+      const out = await this.daico.withdraw({from: owner});
+      const fee = out.receipt.gasUsed / 1000;
+      const final_balance = web3.eth.getBalance(owner).c[0];
+      assert.ok(final_balance - initial_balance - 100 + fee < 1);
```
Comment (facuspagnuolo) [inline test/DAICO.test.js]: we should check `lastWithdrawn` has been successfully updated here too

File: test/DAICO.test.js [hunk 8]
```diff
@@ -33,85 +33,188 @@ contract('DAICO', function (accounts) {
...
+    });
+
+  });
```
Comment (facuspagnuolo) [inline test/DAICO.test.js]: There is a lot repeated code to setup some of these tests, feel free to use a `beforeEach` statement to reduce it

**Review by facuspagnuolo — COMMENTED** [review]
Left some comments!

File: test/DAICO.test.js (hunk 3 above)
Comment (ababino) [reply test/DAICO.test.js]: I mean the period after the `contribution`. I thought I had read the term in Vitalik's original post, but it is not there. Maybe I peaked that up from this [source](https://medium.com/coin-governance-system/daico-and-coin-governance-system-differences-and-similarities-81ba6afc7b9b)

File: test/DAICO.test.js (hunk 8 above)
Comment (ababino) [reply test/DAICO.test.js]: This is the first thing that I did. That is why commit `c185a55` is big, sorry for that.

File: test/DAICO.test.js (hunk 7 above)
Comment (ababino) [reply test/DAICO.test.js]: I did this in `29dfd79` commit

File: test/DAICO.test.js (hunk 6 above)
Comment (ababino) [reply test/DAICO.test.js]: I did this in `29dfd79` commit

File: test/DAICO.test.js (hunk 5 above)
Comment (ababino) [reply test/DAICO.test.js]: I did this in `80b4189` and `6ccb662` commits

File: test/DAICO.test.js (hunk 4 above)
Comment (ababino) [reply test/DAICO.test.js]: I did this in `00663f0` commit

File: test/DAICO.test.js (hunk 2 above)
Comment (ababino) [reply test/DAICO.test.js]: done

File: test/DAICO.test.js [hunk 9]
```diff
@@ -31,87 +31,214 @@ contract('DAICO', function (accounts) {
 
   beforeEach(async function () {
     this.lastWithdrawn = latestTime() + duration.weeks(1);
-    this.afterlastWithdrawn = this.lastWithdrawn + duration.weeks(1)
+    this.afterlastWithdrawn = this.lastWithdrawn + duration.weeks(1);
+    this.owner = accounts[0];
+    this.no_owner = accounts[1];
+    this.purchaser = accounts[2];
+    this.no_purchaser = accounts[3];
```
Comment (facuspagnuolo) [inline test/DAICO.test.js:38]: not a blocker, you can define all these variables within the `contract()` statement as follows:
```
contract('DAICO', function([owner, no_owner, purchaser, no_purchaser]) {...
```

**Review by facuspagnuolo — APPROVED** [review]
Let's merge it, I would like to take another look to all the test suite once we divide the voting process from the DAICO contract

File: test/DAICO.test.js (hunk 9 above)
Comment (ababino) [reply test/DAICO.test.js:38]: great!

**Pull request merged**